src/providers/ogr/qgsogrfeatureiterator.cpp

```diff
@@ -282,20 +296,38 @@ bool QgsOgrFeatureIterator::fetchFeature( QgsFeature &feature )
 
   gdal::ogr_feature_unique_ptr fet;
 
-  while ( fet.reset( OGR_L_GetNextFeature( mOgrLayer ) ), fet )
+  // OSM layers (especially large ones) need the GDALDataset::GetNextFeature() call rather than OGRLayer::GetNextFeature()
+  // see more details here: https://trac.osgeo.org/gdal/wiki/rfc66_randomlayerreadwrite
+  if ( mSource->mDriverName == QLatin1String( "OSM" ) ) 
```

Comment:
this should be protected by #ifdef GDAL_VERSION_NUM >= GDAL_COMPUTE_VERSION(2,2,0)